### PR TITLE
fix: skip fetching checkpoint from rpc nodes when encounter extended primary failures

### PR DESCRIPTION
**Pull request by halfprice**
## Description

When an RetriableRpcClient encounters failures for extended period of time when fetching checkpoints from RPC nodes, the node should just skip RPC for short period of time, and directly fetch checkpoints from archival. Otherwise, the RPC request + retry + failure takes 40s to 60s, which significantly limits the checkpoint downloading throughput.

## Test plan

Tested archival catchup in PTN
Screenshot 2025-05-14 at 1 17 00 AM

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.
For each box you select, include information after the relevant heading that describes the impact of your changes that
a user might notice and any actions they must take to implement updates. (Add release notes after the colon for each item)

- [ ] Storage node:
- [ ] Aggregator:
- [ ] Publisher:
- [ ] CLI:

### CODE REVIEW

File: crates/checkpoint-downloader/src/downloader.rs [hunk 1]
```diff
@@ -265,7 +263,6 @@ impl ParallelCheckpointDownloaderInner {
         let mut last_scale = Instant::now();
         let mut consecutive_failures = 0;
         let average_workers = (downloader_config.min_workers + downloader_config.max_workers) / 2;
-        const MAX_CONSECUTIVE_POOL_MONITOR_FAILURES: u32 = 100;
```
Comment (halfprice) [inline crates/checkpoint-downloader/src/downloader.rs:270]: @sadhansood , how was this count originally determined? In PTN, it took nodes more than an hour to reach to 100 failures to check checkpoint lag.

I convert it into a config setting (and a few other parameters), so that we can easily change it.

File: crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs [hunk 2]
```diff
@@ -554,16 +573,9 @@ impl RetriableClientError {
...
             }
-            _ => self.is_failure_window_exceeded(last_success, num_failures),
+            _ => false,
         }
     }
-
-    /// Returns `true` if the failure window has been exceeded. Failure window is exceeded if
-    /// the number of failures exceeds `MAX_FAILURES` and if the last successful RPC call was
-    /// more than `FAILURE_WINDOW` minutes ago.
-    fn is_failure_window_exceeded(&self, last_success: Instant, num_failures: usize) -> bool {
```
Comment (halfprice) [inline crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs:612]: Note: this part is now moved to FallbackClient so that it can be checked against config.

File: crates/checkpoint-downloader/src/downloader.rs [hunk 3]
```diff
@@ -426,75 +424,42 @@ impl ParallelCheckpointDownloaderInner {
...
 
-                let delay = backoff.next().expect("backoff should not be exhausted");
+                if let Some(delay) = backoff.next() {
+                    tokio::time::sleep(delay).await;
+                    continue;
+                }
 
-                tokio::time::sleep(delay).await;
-                continue;
+                return CheckpointEntry::new(sequence_number, Err(err.into()));
```
Comment (sadhansood) [inline crates/checkpoint-downloader/src/downloader.rs]: I don't think this logic should be changed to return an error as that would cause an error to be returned to the client (event processor in this case). The config `min_retries` should be removed imo since it is not being used.

File: crates/checkpoint-downloader/src/downloader.rs (hunk 1 above)
Comment (sadhansood) [reply crates/checkpoint-downloader/src/downloader.rs:270]: I did not want it to start adjusting workers immediately after full nodes go down and ~ 1 hour seems reasonable

File: crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs [hunk 4]
```diff
@@ -293,58 +301,80 @@ impl RetriableRpcClient {
...
-        if !error.is_eligible_for_fallback(
-            sequence_number,
-            self.last_success.load(Ordering::Relaxed),
-            self.num_failures.load(Ordering::Relaxed),
-        ) {
-            tracing::debug!(
-                "primary client error while fetching checkpoint is not eligible for fallback"
+            // RPC node failure is sustained, so we skip it for 5 minutes.
+            // Also, at this point, the fallback client must be set.
+            self.skip_rpc_for_checkpoint_until.store(
```
Comment (sadhansood) [inline crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs:422]: Does this also get set for something like missing events error? Since it is in the check for `is_elligible_for_fallback`

File: crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs [hunk 5]
```diff
@@ -293,58 +301,80 @@ impl RetriableRpcClient {
...
         let _scope = monitored_scope("RetriableRpcClient::get_full_checkpoint");
         let start_time = Instant::now();
-        let error = match self.get_full_checkpoint_from_primary(sequence_number).await {
-            Ok(checkpoint) => {
+
+        // Check if we should directly skip RPC node due to previous failures, and use fallback.
+        let try_rpc_node_first = self.fallback_client.is_none()
+            || start_time >= self.skip_rpc_for_checkpoint_until.load(Ordering::Relaxed);
+
+        if try_rpc_node_first {
```
Comment (sadhansood) [inline crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs:370]: It'll be nice to refactor some of the logic in a separate function, it seems like this function is getting bigger and harder to follow. Also, could we update the docstring of this function?

**Review by sadhansood — COMMENTED** [review]
Thanks @halfprice , the PR looks solid, I just have a few comments for your consideration.

File: crates/checkpoint-downloader/src/downloader.rs (hunk 3 above)
Comment (halfprice) [reply crates/checkpoint-downloader/src/downloader.rs]: The code used to have two `download_with_retry` function, which was confusing sometimes. The change here just merged these two together, so that we don't have one that does retry and one don't.

I made some changes in the code to indicate that the error returned here is only possible in tests, which just to make `test_parallel_fetcher` work. PTAL.

File: crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs (hunk 4 above)
Comment (halfprice) [reply crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs:422]: Great catch! We should only skip calling RPC node when `is_failure_window_exceeded`. (one thing I don't feel to comfortable about this PR is that, for realistic scenario like this, it's very hard to test/discover such corner cases).

File: crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs (hunk 5 above)
Comment (halfprice) [reply crates/walrus-sui/src/client/retry_client/retriable_rpc_client.rs:370]: I made a pass, and hopefully it's cleaner now. PTAL.

File: crates/checkpoint-downloader/src/downloader.rs [hunk 6]
```diff
@@ -426,75 +424,54 @@ impl ParallelCheckpointDownloaderInner {
...
     /// Downloads a checkpoint with retries.
-    #[cfg(not(test))]
     async fn download_with_retry(
         client: &RetriableRpcClient,
         sequence_number: CheckpointSequenceNumber,
         config: &ParallelDownloaderConfig,
         rng: &mut StdRng,
     ) -> CheckpointEntry {
-        let mut backoff = create_backoff(rng, config);
+        let mut backoff = if cfg!(test) {
```
Comment (sadhansood) [inline crates/checkpoint-downloader/src/downloader.rs:435]: nit: Maybe add a comment about the different behavior b/w prod and test for future. it's a small thing but I think we might scratch our heads one day later and wonder why this if/else branch is needed.

**Review by sadhansood — APPROVED** [review]
Thanks @halfprice, this PR is going to be really valuable during fallback scenarios!

File: crates/checkpoint-downloader/src/downloader.rs (hunk 6 above)
Comment (halfprice) [reply crates/checkpoint-downloader/src/downloader.rs:435]: Done.
